diff --git a/core/src/main/scala/kafka/server/KafkaApis.scala b/core/src/main/scala/kafka/server/KafkaApis.scala
index f68fec5695..857e0c2116 100644
--- a/core/src/main/scala/kafka/server/KafkaApis.scala
+++ b/core/src/main/scala/kafka/server/KafkaApis.scala
@@ -88,6 +88,9 @@ import scala.collection.{Map, Seq, Set, immutable, mutable}
 import scala.jdk.CollectionConverters._
 import scala.util.{Failure, Success, Try}
 
+import kafka.metrics.ProducerMetricsCollector
+import org.apache.kafka.common.requests.{ProduceRequest, ProduceResponse}
+
 /**
  * Logic to handle the various Kafka requests
  */
@@ -603,8 +606,10 @@ class KafkaApis(val requestChannel: RequestChannel,
   /**
    * Handle a produce request
    */
+  val metricsCollector = ProducerMetricsCollector.initialize(config)
   def handleProduceRequest(request: RequestChannel.Request, requestLocal: RequestLocal): Unit = {
     val produceRequest = request.body[ProduceRequest]
+    metricsCollector.interceptProduceRequest(request.header.clientId, request.context.clientAddress.getHostAddress, produceRequest)
 
     if (RequestUtils.hasTransactionalRecords(produceRequest)) {
       val isAuthorizedTransactional = produceRequest.transactionalId != null &&
@@ -697,6 +702,7 @@ class KafkaApis(val requestChannel: RequestChannel,
         }
       }
 
+      metricsCollector.interceptProduceResponse(request.header.clientId, request.context.clientAddress.getHostAddress, mergedResponseStatus.toMap)
       // Send the response immediately. In case of throttling, the channel has already been muted.
       if (produceRequest.acks == 0) {
         // no operation needed if producer request.required.acks = 0; however, if there is any error in handling
